### PR TITLE
Implemented fidelous encryption algorithm

### PR DESCRIPTION
**Pull request by VenuAjitesh**
Added Encryption algorithm and logic behind encryption is :
- After receiving the bundle from the facility, the controller requests for encryption
- EncryptionController request for keys via KeyController
- Using the keys encryption is processed and returns to the DataTransferInterface

### CODE REVIEW

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
+/* (C) 2024 */
```
Comment (chandrashekar-s) [inline src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:1]: Please add a reference link from where this is copied from

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+import org.bouncycastle.crypto.params.KeyParameter;
+import org.bouncycastle.jce.interfaces.ECPublicKey;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.jce.spec.ECParameterSpec;
+import org.bouncycastle.jce.spec.ECPrivateKeySpec;
+import org.bouncycastle.jce.spec.ECPublicKeySpec;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.RequestBody;
+
+public class EncryptionController {
```
Comment (chandrashekar-s) [inline src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:37]: The name `XXXController` will usually given for classes which exposes Rest APIs, check if thats the case here.

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+  private static final Logger log = LogManager.getLogger(EncryptionController.class);
+  @Autowired KeyController keyController;
+
+  public EncryptionResponse encrypt(
+          @RequestBody HIPHealthInformationRequest hipHealthInformationRequest, BundleResponseHIP bundleResponse)
```
Comment (chandrashekar-s) [inline src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:42]: Why `RequestBody` is needed here?

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java (hunk 2 above)
Comment (chandrashekar-s) [reply src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:37]: Can you also show how is this class used?

**Review by chandrashekar-s — COMMENTED** [review]
Thanks for the changes @VenuAjitesh. Looks like the changes are being copied from sample implementation. Can you add a reference link to the code?

Also, try to show the usage of the classes.

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+      byte[] plainBytes = new byte[cipher.getOutputSize(stringBytes.length)];
+      int retLen = cipher.processBytes(stringBytes, 0, stringBytes.length, plainBytes, 0);
+      cipher.doFinal(plainBytes, retLen);
+
+      encryptedData = getBase64String(plainBytes);
+    } catch (Exception e) {
+      System.out.println(e.getLocalizedMessage());
+    }
+
+    log.info("EncryptedData: " + encryptedData);
```
Comment (atulai-sg) [inline src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:110]: lets add these as debug statements

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java (hunk 2 above)
Comment (VenuAjitesh) [reply src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:37]: All these encryption related code are being used in the latest PR, when we get the request for data transfer I'm calling the  encryption controller for encrypting the bundle

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java (hunk 3 above)
Comment (VenuAjitesh) [reply src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:42]: For encryption, i need the bundle response from the facility and also the receiver public keys from HIPHealthInformationRequest to encrypt the FHIR bundle

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  public byte[] getBytesForBase64String(String value) {
+    return org.bouncycastle.util.encoders.Base64.decode(value);
+  }
+
+  public String encrypt(
+      byte[] xorOfRandom, String senderPrivateKey, String receiverPublicKey, String stringToEncrypt)
+      throws Exception {
+    System.out.println("<------------------- ENCRYPTION -------------------->");
```
[the remark this answers is not in the document]
Comment (VenuAjitesh) [reply src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:82]: These have been removed in the latest PR

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java (hunk 4 above)
Comment (VenuAjitesh) [reply src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:110]: Sure they can be modified.

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java (hunk 3 above)
Comment (VenuAjitesh) [reply src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:42]: Oh, I get that, it is not needed here.

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java (hunk 1 above)
Comment (VenuAjitesh) [reply src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:1]: https://github.com/sukreet/fidelius

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java (hunk 2 above)
Comment (VenuAjitesh) [reply src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:37]: And no, we are not exposing any REST API so can you suggest a better name..?

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java (hunk 1 above)
Comment (chandrashekar-s) [reply src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:1]: Add this link in the `EncrytionController` class in the class level javadoc

File: src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java (hunk 2 above)
Comment (chandrashekar-s) [reply src/main/java/com/nha/abdm/wrapper/common/dataPackaging/encryption/EncryptionController.java:37]: Ok. I think @atulai-sg has suggested some names in his comments. Please refer that.

**Pull request merged**